### DISCUSSION TITLE
How can I change image registry?

### DISCUSSION THREAD

**Discussion by sky96111** [home-assistant/operating-system#2797] [Q&A] [answered]
In China ghcr.io is not accessable.
like this reply https://github.com/home-assistant/operating-system/issues/2135#issuecomment-1457402856. We can hardly init or upgrade HAOS.

on Homeassistant Core, we can simply modify
``docker pull ghcr.io/home-assistant/home-assistant:latest``

to ``docker pull docker.io/homeassistant/home-assistant:latest`` (official images on docker hub)
or ``docker pull ghcr.nju.edu.cn/home-assistant/home-assistant:latest`` (an ghcr.io mirror by Nanjing University)

But methods above is not available on HAOS.
Is there any way to edit image registry on HAOS? It would be better if the modification can be saved during upgrade.
Thanks!

**Comment by black-roland** [comment, marked as answer]
@sky96111 @yiqian987 Recently I found a hack to override configuration in `/etc/docker/daemon.json` but I'm not sure if this is relevant to your case. Through `daemon.json` it's possible to set docker.nju.edu.cn as registry mirror but that wouldn't work for the ghcr.io registry (only for the primary one at hub.docker.com).

`daemon.json` also allows you to configure a proxy server. Are there any proxies in China that you can use to access ghcr.io?

Anyway, the hack idea is to use an udev rule to mount `daemon.json` on top of the read-only one. In HAOS you can configure [custom udev rules using a USB stick](https://github.com/home-assistant/operating-system/blob/dev/Documentation/configuration.md) so you need to create 2 files on the USB stick (or inside `CONFIG`):

<details>
<summary markdown="span">udev/00-docker-mirrors-workaround.rules</summary>

```
ENV{ID_FS_LABEL}="hassos-overlay", ACTION=="change", RUN+="/usr/bin/systemd-mount --no-block -o bind /etc/udev/rules.d/docker-daemon.json /etc/docker/daemon.json"
```
</details>

<details>
<summary markdown="span">udev/docker-daemon.json</summary>

```json
{
    "storage-driver": "overlay2",
    "log-driver": "journald",
    "experimental": true,
    "ip6tables": true,
    "log-opts": {
        "tag": "{{.Name}}"
    },
    "data-root": "/mnt/data/docker",
    "deprecated-key-path": "/mnt/overlay/etc/docker/key.json",
    "bip": "172.30.232.1/23",
    "registry-mirrors": [
        "https://docker.nju.edu.cn/"
    ],
    "proxies": {
        "http-proxy": "http://any.proxy.with.access.to.ghcr.io:3128",
        "https-proxy": "https://any.proxy.with.access.to.ghcr.io:3129",
        "no-proxy": "*.test.example.com,.example.org,127.0.0.0/8"
    }
}
```
</details>

Docker will be using this new configuration after system reboot.

I'm not in China so I'm not able to test this workaround but hopefully this would inspire you to make some working solution :slightly_smiling_face: 

P.S. [Here is a blog post](https://mansmarthome.info/posts/home-assistant/obkhod-blokirovki-docker-hub/) describing the hack in more detail but it's in Russian…

**Reply by sky96111** [reply]
Docker mirror could potentially be a solution, but it doesn't support registry except dockerhub.
I Think set proxy for docker using udev rule should work. I will try it later

**Reply by judgeaxl** [reply]
As far as I can tell this solution no longer works in 2026.2.1 (and perhaps earlier). Some HAOS supported way of configuring respository-mirrors would be much appreciated. I've tried all sorts of ways to route the VM network through external proxies or always-on-VPN networks, but I can't get nothing to work, and none of the methods seem to support only redirecting docker anyway.

**Reply by nicholaelaw** [reply]
> As far as I can tell this solution no longer works in 2026.2.1 (and perhaps earlier). Some HAOS supported way of configuring respository-mirrors would be much appreciated. I've tried all sorts of ways to route the VM network through external proxies or always-on-VPN networks, but I can't get nothing to work, and none of the methods seem to support only redirecting docker anyway.

Tried this on 2016.5.2, the `udev/docker-daemon.json` file if unchanged will cause boot loops. I deleted `storage-driver`, `experimental`, `ip6tables` and `deprecated-key-path` lines and it worked. I can thus confirm this method is still valid. Although that doesn't change the fact that Hass has continued to ignore such a fundamental and simple issue...

**Reply by nicholaelaw** [reply]
Also, you can put these 2 files in `/etc/udev/rules.d/`, instead of on a USB stick.
